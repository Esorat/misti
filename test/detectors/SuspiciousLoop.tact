const TRUE: Bool = true;
const FALSE: Bool = false;
const A: Int = 1000;
const B: Int = 10000;

contract TestSuspiciousLoops {

    fun testWhileInfinite() {
        let i: Int = 0;
        while (true) { i = i + 1; } // Bad
        do { i = i + 1; } until (true); // Bad
        do { i = i + 1; } until (TRUE); // TODO: Unsupported in consteval
    }

    fun testRepeatHighCount() {
        repeat (10_001) { let x = 1; } // Bad
        repeat (A + B) { let x = 1; } // TODO: Unsupported in consteval
        repeat (A) { let x = 1; } // OK
    }

    fun testNestedLoops() {
        let i: Int = 0;
        while (i < 10) {
            repeat (1_000_000) { i = i + 1; } // Bad
        }
    }

    fun testLoop() {
        let a: Int = 0;
        while (false) { a = 1; } // Bad
        while (false) { } // OK: Empty body
        do { a = 1; } until (false); // OK: Until is allowed
        do {} until (false); // OK: Until is allowed
        while (TRUE && FALSE) { a = 1; } // TODO: Unsupported in consteval
    }
}
