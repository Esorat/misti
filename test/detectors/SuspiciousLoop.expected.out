[HIGH] SuspiciousLoop: Potential infinite loop
test/detectors/SuspiciousLoop.tact:5:16:
  4 |         let i: Int = 0;
> 5 |         while (true) { // This should be flagged for an unbounded condition
                     ^
  6 |             i = i + 1;
Help: Avoid unbounded conditions in loops
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousLoop

[HIGH] SuspiciousLoop: Potential high-cost loop
test/detectors/SuspiciousLoop.tact:11:17:
  10 |     fun testRepeatHighCount() { 
> 11 |         repeat (1_000_001) { // This should be flagged for excessive iteration
                       ^
  12 |             let x = 1;
Help: Avoid excessive iterations in loops
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousLoop

[HIGH] SuspiciousLoop: Potential high-cost loop
test/detectors/SuspiciousLoop.tact:19:21:
  18 |         while (i < 10) {
> 19 |             repeat (1_000_000) { // Should be flagged within nested loop
                           ^
  20 |                 i = i + 1;
Help: Avoid excessive iterations in loops
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousLoop

[HIGH] SuspiciousLoop: Loop condition is always false
test/detectors/SuspiciousLoop.tact:27:12:
  26 |     let a: Int = 0;
> 27 |     while (false) { // Should be flagged for dead code
                  ^
  28 |         a = 10000000; true; 
Help: The condition is always false; therefore, the body will never execute
See: https://nowarp.io/tools/misti/docs/detectors/SuspiciousLoop